File: RailgunVB/Core/Configuration/MasterConfig.vb
Imports System.IO
Imports Newtonsoft.Json
Imports RailgunVB.Core.Logging

Namespace Core.Configuration
    
    <JsonObject(MemberSerialization.Fields)>
    Public Class MasterConfig
        <JsonIgnore> Private Const Filename As String = "masterconfig.json"
        
        Public ReadOnly Property DiscordConfig As DiscordConfig
        Public ReadOnly Property PostgreDatabaseConfig As DatabaseConfig
        Public ReadOnly Property MongoDatabaseConfig As DatabaseConfig
        
        Public ReadOnly Property GoogleApiToken As String
        Public ReadOnly Property RandomCatApiToken As String

        <JsonConstructor>
        Public Sub New (
            discordConfig As DiscordConfig,
            postgreDatabaseConfig As DatabaseConfig,
            mongoDatabaseConfig As DatabaseConfig,
            googleApiToken As String,
            randomCatApiToken As String
        )
            Me.DiscordConfig = discordConfig
            Me.PostgreDatabaseConfig = postgreDatabaseConfig
            Me.MongoDatabaseConfig = mongoDatabaseConfig
            Me.GoogleApiToken = googleApiToken
            Me.RandomCatApiToken = randomCatApiToken
        End Sub
        
        Public Shared Async Function LoadAsync() As Task(Of MasterConfig)
            If Not (File.Exists(Filename)) Then Return Await SetupAsync()
            
            Dim json As String = Await File.ReadAllTextAsync(Filename)
            Return JsonConvert.DeserializeObject(Of MasterConfig)(json)
        End Function
        
        Private Shared Async Function SetupAsync() As Task(Of MasterConfig)
            ' <===== DISCORD CONFIGURATION SETUP =====>
            Dim token As String = SetupInput("Discord || Token: ")
            Dim prefix As String = SetupInput("Discord || Prefix [!]: ", "!")
            Dim masterAdminId As ULong = ULong.Parse(SetupInput("Discord || Master Admin ID: "))
            
            Dim discordConfig As New DiscordConfig(token, prefix, masterAdminId)
            ' <=======================================>
            
            ' <===== POSTGRE DATABASE CONFIGURATION SETUP =====>
            Dim phostname As String = SetupInput("Database || Postgre || Hostname [localhost]: ", "localhost")
            Dim pusername As String = SetupInput("Database || Postgre || Username: ")
            Dim ppassword As String = SetupInput("Database || Postgre || Password: ")
            Dim pdatabase As String = SetupInput("Database || Postgre || Database: ")
            
            Dim postgreDatabaseConfig as New DatabaseConfig(phostname, pusername, ppassword, pdatabase)
            ' <==============================================>
            
            ' <===== MONGO DATABASE CONFIGURATION SETUP =====>
            Dim mhostname As String = SetupInput("Database || Mongo || Hostname [localhost]: ", "localhost")
            Dim musername As String = SetupInput("Database || Mongo || Username: ")
            Dim mpassword As String = SetupInput("Database || Mongo || Password: ")
            
            Dim mongoDatabaseConfig as New DatabaseConfig(mhostname, musername, mpassword, Nothing)
            ' <==============================================>

            Dim googleApiKey As String = SetupInput("Other || Google API Key: ")
            Dim randomCatApiKey As String = SetupInput("Other || RandomCat API Key: ")

            Dim masterConfig As New MasterConfig(discordConfig, postgreDatabaseConfig, mongoDatabaseConfig, googleApiKey, randomCatApiKey)
            await masterConfig.SaveAsync()
            Return masterConfig
        End Function
        
        Private Shared Function SetupInput(query As String) As String
            Console.Write(query)
            Return Console.ReadLine()
        End Function
        
        Private Shared Function SetupInput(query As String, defaultTo As String) As String
            Console.Write(query)
            Dim output As String = Console.ReadLine()
            Return If(String.IsNullOrWhiteSpace(query), defaultTo, output)
        End Function
        
        Private Async Function SaveAsync() As Task
            Await File.WriteAllTextAsync(Filename, JsonConvert.SerializeObject(Me, Formatting.Indented))
        End Function
        
        Public Async Function AssignAsMasterGuildAsync(guildId As ULong) As Task
            _DiscordConfig.AssignMasterGuild(guildId)
            Await SaveAsync()
        End Function
        
        Public Async Function AssignBotLogChannelAsync(channelId As ULong, botLogType As BotLogType) As Task
            _DiscordConfig.AssignBotLogChannel(channelId, botLogType)
            Await SaveAsync()
        End Function
        
        Public Async Function AssignPrefixAsync(input As String) As Task
            DiscordConfig.Prefix = input
            Await SaveAsync()
        End Function
        
        Public Async Function AssignAdminAsync(userId As ULong) As Task(Of Boolean)
            If Not (_DiscordConfig.AssignAdmin(userId)) Then Return False
            
            Await SaveAsync()
            Return True
        End Function
        
        Public Async Function RemoveAdminAsync(userId As ULong) As Task(Of Boolean)
            If Not (_DiscordConfig.RemoveAdmin(userId)) Then Return False
            
            Await SaveAsync()
            Return True
        End Function
        
    End Class
    
End Namespace

File: Railgun/Core/Configuration/MasterConfig.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using Newtonsoft.Json;
using Railgun.Core.Enums;

namespace Railgun.Core.Configuration
{
    [JsonObject(MemberSerialization.Fields)]
    public class MasterConfig
    {
        [JsonIgnore]
        private const string Filename = "masterconfig.json";
        [JsonIgnore]
        public string YoutubeDirectory { get; } = "/";
        [JsonIgnore]
        public string YoutubeDomain { get; } = "youtube.com";

        public DiscordConfig DiscordConfig { get; private set; }
        public DatabaseConfig PostgreSqlConfig { get; private set; }
        public DatabaseConfig MongoDbConfig { get; private set; }

        public Dictionary<string, string> YoutubeCookies { get; private set; } = new Dictionary<string, string>();

        public string GoogleApiToken { get; private set; }
        public string RandomCatApiToken { get; private set; }

        public MasterConfig(DiscordConfig discordConfig, DatabaseConfig postgresql, DatabaseConfig mongodb, string googleApiToken, string randomCatApiKey)
        {
            DiscordConfig = discordConfig;
            PostgreSqlConfig = postgresql;
            MongoDbConfig = mongodb;
            GoogleApiToken = googleApiToken;
            RandomCatApiToken = randomCatApiKey;
        }

        [JsonConstructor]
        private MasterConfig() {}

        private static string SetupInput(string query)
        {
            Console.Write(query);
            return Console.ReadLine();
        }

        private static string SetupInput(string query, string defaultTo)
        {
            Console.Write(query);
            var output = Console.ReadLine();
            return string.IsNullOrWhiteSpace(output) ? defaultTo: output;
        }

        public static MasterConfig Load() 
        {
            if (!File.Exists(Filename)) return Setup();
            return JsonConvert.DeserializeObject<MasterConfig>(File.ReadAllText(Filename));
        }

        private static MasterConfig Setup()
        {
            var token = SetupInput("Discord || Token: ");
            var prefix = SetupInput("Discord || Prefix [!]: ", "!");
            var masterAdminId = ulong.Parse(SetupInput("Discord || Master Admin ID: "));
            var discordConfig = new DiscordConfig(token, prefix, masterAdminId);

            var phostname = SetupInput("Database || Postgre || Hostname [localhost]: ", "localhost");
            var pusername = SetupInput("Database || Postgre || Username: ");
            var ppassword = SetupInput("Database || Postgre || Password: ");
            var pdatabase = SetupInput("Database || Postgre || Database: ");
            var postgresqlConfig = new DatabaseConfig(phostname, pusername, ppassword, pdatabase);

            var mhostname = SetupInput("Database || Mongo || Hostname [localhost]: ", "localhost");
            var musername = SetupInput("Database || Mongo || Username: ");
            var mpassword = SetupInput("Database || Mongo || Password: ");
            var mongodbConfig = new DatabaseConfig(mhostname, musername, mpassword, null);

            var googleApiKey = SetupInput("Other || Google API Key: ");
            var randomCatApiKey = SetupInput("Other || RandomCat API Key: ");

            var masterConfig = new MasterConfig(discordConfig, postgresqlConfig, mongodbConfig, googleApiKey, randomCatApiKey);
            masterConfig.Save();

            return masterConfig;
        }

        private void Save()
            => File.WriteAllText(Filename, JsonConvert.SerializeObject(this, Formatting.Indented));
        
        public void AssignMasterGuild(ulong guildId)
        {
            DiscordConfig.AssignMasterGuild(guildId);
            Save();
        }

        public void AssignBotLogChannel(ulong channelId, BotLogType logType)
        {
            DiscordConfig.AssignBotLogChannel(channelId, logType);
            Save();
        }

        public void AssignPrefix(string prefix)
        {
            DiscordConfig.Prefix = prefix;
            Save();
        }

        public bool AssignAdmin(ulong userId)
        {
            if (!DiscordConfig.AssignAdmin(userId)) return false;
            Save();
            return true;
        }

        public bool RemoveAdmin(ulong userId)
        {
            if (!DiscordConfig.RemoveAdmin(userId)) return false;
            Save();
            return true;
        }

        public void UpdateYoutubeCookies(IEnumerable<Cookie> cookies)
        {
            var newCookies = new Dictionary<string, string>();

            foreach (var cookie in cookies)
            {
                if (newCookies.ContainsKey(cookie.Name))
                {
                    newCookies[cookie.Name] = cookie.Value;
                    continue;
                }

                newCookies.Add(cookie.Name, cookie.Value);
            }

            YoutubeCookies = newCookies;
            Save();
        }
    }
}